gateway/src/metrics.rs

```diff
@@ -0,0 +1,1069 @@
+        }
+
+        if dropped_chunks > 0 {
+            slog::info!(
+                &self.log,
+                "produced metric samples. some old sample chunks were dropped!";
+                "samples" => total_samples,
+                "sample_chunks" => samples.len(),
+                "dropped_chunks" => dropped_chunks,
+            );
+        } else {
+            slog::debug!(
+                &self.log,
+                "produced metric samples";
+                "samples" => total_samples,
+                "sample_chunks" => samples.len(),
+            );
+        }
+
+        // There you go, that's all I've got.
+        Ok(Box::new(samples.into_iter().flatten()))
+    }
+}
+
+impl PollerManager {
+    async fn run(
+        mut self,
+        rack_id: oneshot::Receiver<Uuid>,
+    ) -> anyhow::Result<()> {
+        let switch = &self.apictx.mgmt_switch;
+
+        // First, wait until we know what the rack ID is...
+        let rack_id = rack_id.await.context(
+            "rack ID sender has gone away...we must be shutting down",
+        )?;
+
+        let mut known_sps: HashMap<SpIdentifier, tokio::task::AbortHandle> =
+            HashMap::with_capacity(NORMAL_NUMBER_OF_SPS);
+        // Wait for SP discovery to complete, if it hasn't already.
+        // TODO(eliza): presently, we busy-poll here. It would be nicer to
+        // replace the `OnceLock<Result<LocationMap, ...>` in `ManagementSwitch`
+        // with a `tokio::sync::watch`
+        backoff::retry_notify_ext(
+            backoff::retry_policy_local(),
+            || async {
+                if switch.is_discovery_complete() {
+                    Ok(())
+                } else {
+                    Err(backoff::BackoffError::transient(()))
+                }
+            },
+            |_, _, elapsed| {
+                let secs = elapsed.as_secs();
+                if secs < 30 {
+                    slog::debug!(
+                        &self.log,
+                        "waiting for SP discovery to complete...";
+                        "elapsed" => ?elapsed,
+                    );
+                } else if secs < 180 {
+                    slog::info!(
+                        &self.log,
+                        "still waiting for SP discovery to complete...";
+                        "elapsed" => ?elapsed,
+                    )
+                } else {
+                    slog::warn!(
+                        &self.log,
+                        "we have been waiting for SP discovery to complete \
+                         for a pretty long time!";
+                        "elapsed" => ?elapsed,
+                    )
+                }
+            },
+        )
+        .await
+        .expect("we should never return a fatal error here");
+
+        slog::info!(
+            &self.log,
+            "starting to polling SP sensor data every {:?}", self.poll_interval;
+        );
+
+        loop {
+            let sps = backoff::retry_notify_ext(
```

Comment:
I think this could be lifted out of the `loop`; once `all_sps()` returns successfully, it will never change (a change would require a config change and MGS restart). It being in the loop seems a little suspect when combined with the `// All pollers started! Now wait to see if any of them have died...` below, because we'd be waiting on a task to die before we check for any new SPs here.

Reply:
Ah, I hadn't realized that. This makes the implementation a bit simpler.

Reply:
My original code here was based on the  misunderstanding that the state of the `LocationMap` would ever change at runtime, as new SPs are detected, but I've since realized that there will always be an entry for every slot, and the `SingleSp` type is what has the `watch` channel indicating whether the SP is present or not. So, I think the `PollerManager` task is not really necessary: we can just start a poller for every slot, and have them never return an error, and just fall back to waiting for the SP to exist  if it doesn't already.

Reply:
Yeah, it's possible some of the names here (particularly `SingleSp`) are misleading.

* `LocationMap` is the runtime version of the [`[[switch.port]]`](https://github.com/oxidecomputer/omicron/blob/f1e1aff423fd309cb7cbd4994be599c16c7351b2/smf/mgs/config.toml#L68) section of MGS's config, once we've talked to an SP that determines which switch we're connected to. So once `LocationMap` exists, it has an entry for every port listed in the config, regardless of whether there's an SP actually present.
* `SingleSp` is really a wrapper around a _switch port_ (or more specifically, the interface name inside the switch zone that maps to a port on the VSC7448 switch), not a connection to an SP. So a given `SingleSp` may over time have no SP it can talk to, one SP it can talk to (that may change versions over time), or even multiple SPs (not concurrently) if a sled is swapped out while MGS is running.

Reply:
> * `SingleSp` is really a wrapper around a _switch port_ (or more specifically, the interface name inside the switch zone that maps to a port on the VSC7448 switch), not a connection to an SP. So a given `SingleSp` may over time have no SP it can talk to, one SP it can talk to (that may change versions over time), or even multiple SPs (not concurrently) if a sled is swapped out while MGS is running.

Yeah, this was the bit that was unclear to me --- I had thought `SingleSp` represented a known SP was at the other end of the network connection and that it would become invalidated/unhappy if the SP went away or changed. Calling it something different could be a good idea, but I also think that my change to the error types in #6413 would also have made it clearer what the `ManagementSwitch::sp` method's failure modes are, which is probably sufficient IMO.